chain/client/src/stateless_validation/state_witness_distribution_actions.rs

```diff
@@ -1,29 +1,101 @@
+use std::sync::Arc;
+
 use near_async::messaging::CanSend;
+use near_async::time::Clock;
 use near_chain::Error;
 use near_network::types::{NetworkRequests, PeerManagerAdapter, PeerManagerMessageRequest};
+use near_primitives::stateless_validation::{
+    ChunkStateWitness, ChunkStateWitnessAck, EncodedChunkStateWitness,
+    SignedEncodedChunkStateWitness,
+};
+use near_primitives::validator_signer::ValidatorSigner;
+
+use crate::metrics;
 
 use super::state_witness_distribution_actor::DistributeChunkStateWitnessRequest;
+use super::state_witness_tracker::ChunkStateWitnessTracker;
 
-#[derive(Clone)]
 pub struct StateWitnessDistributionActions {
+    /// Adapter to send messages to the network.
     network_adapter: PeerManagerAdapter,
+    /// Validator signer to sign the state witness.
+    my_signer: Arc<dyn ValidatorSigner>,
+    /// Tracks a collection of state witnesses sent from chunk producers to chunk validators.
+    state_witness_tracker: ChunkStateWitnessTracker,
 }
 
 impl StateWitnessDistributionActions {
-    pub fn new(network_adapter: PeerManagerAdapter) -> Self {
-        Self { network_adapter }
+    pub fn new(
+        clock: Clock,
+        network_adapter: PeerManagerAdapter,
+        my_signer: Arc<dyn ValidatorSigner>,
+    ) -> Self {
+        Self {
+            network_adapter,
+            my_signer,
+            state_witness_tracker: ChunkStateWitnessTracker::new(clock),
+        }
     }
 
     pub fn handle_distribute_chunk_state_witness_request(
-        &self,
+        &mut self,
         msg: DistributeChunkStateWitnessRequest,
     ) -> Result<(), Error> {
-        let DistributeChunkStateWitnessRequest { chunk_validators, signed_witness } = msg;
+        let DistributeChunkStateWitnessRequest { chunk_validators, state_witness } = msg;
+
+        let signed_witness = create_signed_witness(&state_witness, self.my_signer.as_ref())?;
```

Comment:
main changes here, basically just moving `record_witness_sent` and `handle_chunk_state_witness_ack`